go/consensus/cometbft/light/internal.go

```diff
@@ -27,19 +27,61 @@ type lightClient struct {
 	tmc *cmtlight.Client
 }
 
+func tryProviders[R any](
+	ctx context.Context,
+	lc *lightClient,
+	fn func(api.Provider) (R, rpc.PeerFeedback, error),
+) (R, rpc.PeerFeedback, error) {
+	// Primary provider.
+	providers := append([]api.Provider{}, lc.tmc.Primary().(api.Provider))
+	// Additional providers.
+	for _, provider := range lc.tmc.Witnesses() {
+		providers = append(providers, provider.(api.Provider))
+	}
+
+	var (
+		result R
+		err    error
+	)
+	for _, provider := range providers {
+		if ctx.Err() != nil {
+			return result, nil, ctx.Err()
+		}
+
+		var pf rpc.PeerFeedback
+		result, pf, err = fn(provider)
+		if err == nil {
+			return result, pf, nil
+		}
+	}
+
+	// Trigger primary provider refresh if everything fails.
+	providers[0].RefreshPeer()
```

Comment:
Any reasons to not trigger the refresh also on the witnesses?

Reply:
Not sure why this is useful. If remote nodes don't have blocks, their p2p rating will never decrease. Therefore, the peer manager will always serve nodes in the same order, except the first 5, which are shuffled. Since witnesses use 2 nodes, this refresh actually chooses 1 node out of 3.